src/combinatorics/simple_graph/ends/defs.lean

```diff
@@ -0,0 +1,280 @@
+
+/-- The connected component of `v`. -/
+@[reducible] def of_vertex (G : simple_graph V) {v : V} (vK : v ∈ K ᶜ) : G.comp_out K :=
+  connected_component_mk (G.out K) ⟨v, vK⟩
+
+lemma of_vertex_mem (G : simple_graph V) {v : V} (vK : v ∈ K ᶜ) :
+  v ∈ (comp_out.of_vertex G vK : G.comp_out K) := ⟨vK, rfl⟩
+
+lemma of_vertex_eq_of_adj (G : simple_graph V) {v w : V} (vK : v ∈ K ᶜ) (wK : w ∈ K ᶜ) :
+  G.adj v w → comp_out.of_vertex G vK = comp_out.of_vertex G wK :=
+by { rw [connected_component.eq], rintro a, apply adj.reachable, exact a }
+
+/-- The induced subgraph of `V` given by the vertices of `C`. -/
+@[reducible, protected]
+def subgraph (C : comp_out G K) : G.subgraph := (⊤ : G.subgraph).induce (C : set V)
+
+/-- The infinite connected components. -/
+@[reducible]
+def inf (C : G.comp_out K) := (C : set V).infinite
+
+/-- The finite connected components -/
+@[reducible, protected]
+def fin (C : G.comp_out K) := (C : set V).finite
+
+lemma coe_inj {C D : G.comp_out K} : (C : set V) = (D : set V) ↔ C = D := set_like.coe_set_eq
+
+@[simp] lemma nempty (C : G.comp_out K) : (C : set V).nonempty :=
+begin
+  refine C.ind (λ v, _),
+  use v,
+  simp only [set.mem_compl_iff, set_like.mem_coe, mem_supp_iff, connected_component.eq,
+             subtype.coe_eta, exists_prop],
+  exact ⟨v.prop, reachable.refl _⟩,
+end
+
+@[protected]
+lemma outside (C : G.comp_out K) : disjoint K C :=
+begin
+  rw set.disjoint_iff,
+  rintro v ⟨vK, vC⟩,
+  simp only [mem_supp_iff, set.mem_compl_iff, set_like.mem_coe] at vC,
+  exact vC.some vK,
+end
+
+lemma not_mem_of_mem {C : G.comp_out K} {c : V} (cC : c ∈ C) : c ∉ K :=
+λ cK, set.disjoint_iff.mp C.outside ⟨cK, cC⟩
+
+@[protected]
+lemma disjoint (C D : G.comp_out K) (ne : C ≠ D) : disjoint (C : set V) (D : set V) :=
+begin
+  rw set.disjoint_iff,
+  rintro u ⟨uC, uD⟩,
+  simp only [set.mem_compl_iff, set_like.mem_coe, mem_supp_iff] at uC uD,
+  exact ne (uC.some_spec.symm.trans uD.some_spec),
+end
+
+lemma eq_of_not_disjoint (C D : G.comp_out K) (nd : ¬ disjoint (C : set V) (D : set V)) : C = D :=
+begin
+  rw set.not_disjoint_iff at nd,
+  simp only [set_like.mem_coe, mem_supp_iff] at nd,
+  obtain ⟨x, ⟨_, rfl⟩, ⟨_, rfl⟩⟩ := nd, refl,
+end
+
+/--
+No vertex of a component `C` outside of `K` is adjacent to a vertex that is
+neither in `C` nor in `K`.
+-/
+lemma nonadj : ∀ (C : G.comp_out K), ¬ (∃ (c d : V), c ∈ C ∧ d ∉ C ∧ d ∉ K ∧ G.adj c d) :=
+begin
+  refine connected_component.ind _,
+  rintros v ⟨c, d, cC, dnC, dnK, cd⟩,
+  have cd' : (G.out K).reachable (⟨c, not_mem_of_mem cC⟩) ⟨d, dnK⟩ := adj.reachable cd,
+  simp only [set.mem_compl_iff, mem_supp_iff, connected_component.eq, not_exists] at cC dnC,
+  exact dnC dnK (cC.some_spec.symm.trans cd').symm,
+end
+
+/--
+Assuming `G` is preconnected and `K` not empty, given any connected component `C` outside of `K`,
+there exists a vertex `k ∈ K` adjacent to a vertex `v ∈ C`.
+-/
+lemma adj [Gc : G.preconnected] (hK : K.nonempty) :
```

Comment:
This is too ambiguous of a name.

Reply:
`adjacent_right` ?

Reply:
it's now `exists_adj_boundary_pair`. It's not ideal but better than nothing